File: app/src/main/java/ru/abbysoft/wisebuild/model/ComputerPart.java
package ru.abbysoft.wisebuild.model;

import android.graphics.Bitmap;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import ru.abbysoft.wisebuild.databinding.ParamDescription;
import ru.abbysoft.wisebuild.utils.ModelUtils;

/**
 * Abstract computer part component
 *
 * WARNING:
 * All derived part types must have constructor with 0 arguments!
 *
 * @author apopov
 */
public abstract class ComputerPart {

    private static long nextID = 0;

    /**
     * unique part identifier
     */
    private final long id;

    private volatile String name;
    private volatile String trimmedName;
    private volatile String description;
    private volatile int releaseYear;
    private volatile int releaseQuarter;
    private volatile Bitmap photo;
    private volatile long priceUsd;
    private final ComputerPartType type;

    protected ComputerPart(@NonNull ComputerPartType type) {
        this.type = type;
        this.id = nextID++;
    }

    public ArrayList<ParamDescription> getAllParameters() {
        ArrayList<ParamDescription> params = new ArrayList<>(10);

        params.add(ModelUtils.createStringParameter("name"));
        params.add(ModelUtils.createStringParameter("description"));
        params.add(ModelUtils.createLongParameter("priceUsd"));
        params.add(ModelUtils.createIntParameter("releaseYear"));
        params.add(ModelUtils.createIntParameter("releaseQuarter"));

        params.addAll(getTypeParams());

        return params;
    }

    public abstract ArrayList<ParamDescription> getTypeParams();

    public String getName() {
        return name;
    }

    public String getTrimmedName() {
        return trimmedName;
    }

    public void setName(String name) {
        this.name = name;
        this.trimmedName = createTrimmedName();
    }

    private String createTrimmedName() {
        int MAX_TRIMMED_NAME_LENGTH = 20;
        if (name.length() <= MAX_TRIMMED_NAME_LENGTH) {
            return getName();
        }


        String trimmed = name.substring(0, MAX_TRIMMED_NAME_LENGTH);
        trimmed += "...";

        return trimmed;
    }

    public String getDescription() {
        return description;
    }

    public Bitmap getPhoto() {
        return photo;
    }

    public void setPhoto(Bitmap photo) {
        this.photo = photo;
    }

    public long getPriceUsd() {
        return priceUsd;
    }

    public void setPriceUsd(long priceUsd) {
        this.priceUsd = priceUsd;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public ComputerPartType getType() {
        return type;
    }

    public long getId() {
        return id;
    }

    public int getReleaseYear() {
        return releaseYear;
    }

    public void setReleaseYear(int releaseYear) {
        this.releaseYear = releaseYear;
    }

    public int getReleaseQuarter() {
        return releaseQuarter;
    }

    public void setReleaseQuarter(int releaseQuarter) {
        this.releaseQuarter = releaseQuarter;
    }

    public enum ComputerPartType {
        CPU("CPU", ru.abbysoft.wisebuild.model.CPU.class),
        MEMORY_MODULE("Memory module", MemoryModule.class),
        MOTHERBOARD("Motherboard", Motherboard.class),
        ASSEMBLED_PC("Assembled PC", AssembledPC.class);

        private final String readableName;
        private final Class objectClass;

        ComputerPartType(String name, Class objectClass) {
            this.readableName = name;
            this.objectClass = objectClass;
        }

        public static List<ComputerPartType> getEntriesWithoutAssembly() {
            List<ComputerPartType> partTypes = new ArrayList<>(values().length-1);
            for (ComputerPartType type : values()) {
                if (type != ASSEMBLED_PC) {
                    partTypes.add(type);
                }
            }
            return partTypes;
        }

        public String getReadableName() {
            return readableName;
        }

        public Class getObjectClass() {
            return objectClass;
        }
    }
}

File: app/src/main/java/ru/abbysoft/wisebuild/model/ComputerPart.kt
package ru.abbysoft.wisebuild.model

import android.graphics.Bitmap
import ru.abbysoft.wisebuild.databinding.ParamCategory

import java.util.ArrayList

import ru.abbysoft.wisebuild.databinding.ParamDescription
import ru.abbysoft.wisebuild.utils.ModelUtils

/**
 * Abstract computer part component
 *
 * WARNING:
 * All derived part types must have constructor with 0 arguments!
 *
 * @author apopov
 */
abstract class ComputerPart protected constructor(val type: ComputerPartType) {

    /**
     * unique part identifier
     */
    val id: Long
    var name: String? = null
        set(name) {
            field = name
            this.trimmedName = createTrimmedName()
        }
    var trimmedName: String? = null
        private set
    var description: String? = null
    var releaseYear: Int = 0
    var releaseQuarter: Int = 0
    var photo: Bitmap? = null
    var priceUsd: Long = 0

    val allParameters: ArrayList<ParamDescription>
        get() {
            val params = ArrayList<ParamDescription>(10)

            params.add(ModelUtils.createStringParameter("name").apply { required = true })
            params.add(ModelUtils.createStringParameter("description"))
            params.add(ModelUtils.createLongParameter("priceUsd")
                    .apply {
                        paramCategory = ParamCategory.PRICE
                        required = true })
            params.add(ModelUtils.createIntParameter("releaseYear"))
            params.add(ModelUtils.createIntParameter("releaseQuarter"))

            params.addAll(typeParams)

            return params
        }

    abstract val typeParams: ArrayList<ParamDescription>

    init {
        this.id = nextID++

        this.releaseQuarter = -1
        this.releaseYear = -1
    }

    private fun createTrimmedName(): String? {
        val MAX_TRIMMED_NAME_LENGTH = 20
        if (this.name!!.length <= MAX_TRIMMED_NAME_LENGTH) {
            return name
        }


        var trimmed = this.name!!.substring(0, MAX_TRIMMED_NAME_LENGTH)
        trimmed += "..."

        return trimmed
    }

    enum class ComputerPartType constructor(val readableName: String, val objectClass: Class<*>) {
        CPU("CPU", ru.abbysoft.wisebuild.model.CPU::class.java),
        MEMORY_MODULE("Memory module", MemoryModule::class.java),
        MOTHERBOARD("Motherboard", Motherboard::class.java),
        ASSEMBLED_PC("Assembled PC", AssembledPC::class.java),
        VIDEO_CARD("Video card", VideoCard::class.java);


        companion object {
            val entriesWithoutAssembly: List<ComputerPartType>
                get() {
                    val partTypes = ArrayList<ComputerPartType>(values().size - 1)
                    for (type in values()) {
                        if (type != ASSEMBLED_PC) {
                            partTypes.add(type)
                        }
                    }
                    return partTypes
                }
        }
    }

    companion object {

        private var nextID: Long = 0
    }
}
